ffmpeg/lpms_ffmpeg.c

```diff
@@ -102,3 +140,588 @@ int lpms_rtmp2hls(char *listen, char *outf, char *ts_tmpl, char* seg_time)
+    if (ret == AVERROR(EAGAIN)) continue;
+    else if (ret < 0) dec_err("Error receiving frame from decoder\n");
+    break;
+  }
+
+dec_cleanup:
+  if (ret < 0) av_packet_unref(pkt); // XXX necessary? or have caller do it?
+  return ret;
+
+dec_flush:
+  if (ictx->vc) {
+    avcodec_send_packet(ictx->vc, NULL);
+    ret = avcodec_receive_frame(ictx->vc, frame);
+    pkt->stream_index = ictx->vi; // XXX ugly?
+    if (!ret) return ret;
+  }
+  if (ictx->ac) {
+    avcodec_send_packet(ictx->ac, NULL);
+    ret = avcodec_receive_frame(ictx->ac, frame);
+    pkt->stream_index = ictx->ai; // XXX ugly?
+  }
+  return ret;
+
+#undef dec_err
+}
+
+int process_out(struct output_ctx *octx, AVCodecContext *encoder, AVStream *ost,
+  struct filter_ctx *filter, AVFrame *inf)
+{
+#define proc_err(msg) { \
+  char errstr[AV_ERROR_MAX_STRING_SIZE] = {0}; \
+  if (!ret) { fprintf(stderr, "u done messed up\n"); ret = AVERROR(ENOMEM); } \
+  if (ret < -1) av_strerror(ret, errstr, sizeof errstr); \
+  fprintf(stderr, "%s: %s", msg, errstr); \
+  goto proc_cleanup; \
+}
+  int ret = 0;
+  AVFrame *frame = NULL;
+  AVPacket pkt = {0};
+  AVRational tb;
+  if (filter && filter->active && inf) {
+    frame = filter->frame;
+    av_frame_copy_props(frame, inf);
+    ret = av_buffersrc_add_frame_flags(filter->src_ctx, inf, AV_BUFFERSRC_FLAG_KEEP_REF);
+    if (ret < 0) proc_err("Error feeding the filtergraph\n");
+    ret = av_buffersink_get_frame(filter->sink_ctx, frame);
+    frame->pict_type = AV_PICTURE_TYPE_NONE;
+    if (AVERROR(EAGAIN) == ret || AVERROR_EOF == ret) return ret; // XXX do properly
+    if (ret < 0) proc_err("Error consuming the filtergraph");
+    tb = av_buffersink_get_time_base(filter->sink_ctx);
+  } else frame = inf;
+
+  // encode
+  av_init_packet(&pkt);
+  if (encoder) {
+    ret = avcodec_send_frame(encoder, frame);
+    if (AVERROR_EOF == ret) ;
+    else if (ret < 0) proc_err("Error sending frame to encoder\n");
+    ret = avcodec_receive_packet(encoder, &pkt);
+    if (AVERROR(EAGAIN) == ret || AVERROR_EOF == ret) return ret;
+    if (ret < 0) proc_err("Error receiving packet from encoder\n");
+    tb = encoder->time_base;
+  } else proc_err("Trying to transmux") // XXX pass in the inpacket, set  pkt = ipkt
+
+
+  // packet bookkeeping.  XXX use av_rescale_delta for audio
+  pkt.stream_index = ost->index;
+  if (av_cmp_q(tb, ost->time_base)) {
+    pkt.pts = av_rescale_q_rnd(pkt.pts, tb, ost->time_base, AV_ROUND_NEAR_INF | AV_ROUND_PASS_MINMAX);
+    pkt.dts = av_rescale_q_rnd(pkt.dts, tb, ost->time_base, AV_ROUND_NEAR_INF | AV_ROUND_PASS_MINMAX);
+    pkt.duration = av_rescale_q(pkt.duration, encoder->time_base, ost->time_base);
+  }
+
+  ret = av_interleaved_write_frame(octx->oc, &pkt);
+  if (ret < 0) proc_err("Error writing frame\n"); // XXX handle better?
+
+proc_cleanup:
+  av_frame_unref(frame);
+  av_packet_unref(&pkt);
+  return ret;
+#undef proc_err
+}
+
+#define MAX_OUTPUT_SIZE 3
```

Comment:
Why use MAX_OUTPUT_SIZE if we already have nb_outputs?

Reply:
Good catch -- we preallocate MAX_OUTPUT_SIZE output contexts (on the stack), but we're missing the check for nb_outputs <= MAX_OUTPUT_SIZE.